Post-Revolutionary Fervor station trait, revolutionary bedsheets, and a megaphone

Rhials opened the pull request:
## About The Pull Request

Upon revolution success, the chosen headrev will now also receive a megaphone, and a "revolutionary bedsheet" repurposed from a stolen CC bedsheet to commemorate their success. The post-revs confusion and lack of command/security usually leads to an instantaneous, total breakdown in cohesion. It's every man for himself -- that's no way to run a commune! Just because the revolution has succeeded and nobody can see your big blue "R" anymore doesn't mean you can't be a leader! 

![image](https://user-images.githubusercontent.com/28870487/222981576-e62e457b-1b2d-4756-8c87-7a9093c92c2d.png)

This also adds a new revolution-themed negative station trait -- Post-Revolutionary Fervor. When present, this trait trashes the command areas at the start of the round. This means cracked windows, broken consoles, vendors getting knocked over, and the occasional dead greytider.

![image](https://user-images.githubusercontent.com/28870487/222981095-14ce9336-2320-406e-b0a6-dc91cb8f9479.png)

If you start cleaning at the start of the round, you might finish right as the next batch of revs decides to crop up.
## Why It's Good For The Game

Giving one of the headrevs a bigger voice and a cool cape (or uncool, depending on how you view the sprite) means that there's a chance for them to step up and try to keep the wheels on. Just remember -- Nobody is obligated to actually listen to this person, it's just a bedsheet.

Adds a neato station trait, which probably counts as command gameplay content.

## Changelog
:cl: Rhials
add: The headrev who receives the revolutionary banner after a win will also receive a commemorative bedsheet and megaphone.
add: Post-Revolutionary Fervor station trait. I hope you enjoy fixing broken computer screens.
spriteadd: A revolutionary bedsheet.
/:cl:

Rhials committed:
adds it

ahrd to test, not working :(

Rhials committed:
tested and works

need to fix the worn sprites for the bedsheet though

Rhials committed:
inhand sprites

Rhials committed:
adds the mapping helper for rev trashing

untested so far

Rhials committed:
tests out the mapping helper (it works!)

Rhials committed:
NEW TRAIT

Rhials committed:
more stuff, fixes the trait

im gonna test if I can make this just check atoms that would make a lot more sense right

Rhials committed:
strips some fat

Rhials committed:
metastation

Rhials committed:
kilo

Rhials committed:
delta

Rhials committed:
icebox

its like 1 am and i am NOT doing tramstation rn

Rhials committed:
aaaaaaaand tram

Rhials committed:
fixes sheet sprites and rev trait

Rhials committed:
better mapping helper stuff

Rhials committed:
I hope this didnt do anything bad

Rhials committed:
Merge remote-tracking branch 'upstream/master' into revolutionary-bedsheets

Rhials committed:
okay why the fuck is it telling me removed a bunch of shit from bedsheets.dmi

Rhials committed:
removes a comment and some phrasing I didnt like

Rhials committed:
oops

Mothblocks commented on code/modules/mapping/mapping_helpers.dm:
```diff
@@ -851,3 +851,96 @@ INITIALIZE_IMMEDIATE(/obj/effect/mapping_helpers/no_lava)
+
+
+///Causes a room to be trashed and graffitied if the revs won in the last round
+/obj/effect/mapping_helpers/revolution_trash
```
I think we should avoid this pattern as much as possible, especially since we potentially have a new map along the way.

Why are you doing it like this instead of looking for areas?

Mothblocks commented on code/modules/mapping/mapping_helpers.dm:
```diff
@@ -851,3 +851,96 @@ INITIALIZE_IMMEDIATE(/obj/effect/mapping_helpers/no_lava)
+	)
+
+/obj/effect/mapping_helpers/revolution_trash/Initialize(mapload)
+	.=..()
```
```suggestion
	..()
```

Mothblocks commented on code/modules/mapping/mapping_helpers.dm:
```diff
@@ -851,3 +851,96 @@ INITIALIZE_IMMEDIATE(/obj/effect/mapping_helpers/no_lava)
+
+	var/area/our_area = get_area(src)
+
+	for(var/atom/current_thing in our_area.contents)
```
`as anything`

Also, instead of `.contents` (which is *extremely* slow for areas), use the proc we made for this. I think it's `get_area_contents()`? Lemon would know

Mothblocks replied:
I'm not stoked about all the `istype` here. You should probably make a proc for the thrashing instead. I don't know how many areas have this, but given the rarity it's fine performance wise to call a proc on everything in an area.

Mothblocks commented on code/modules/mapping/mapping_helpers.dm:
```diff
@@ -851,3 +851,96 @@ INITIALIZE_IMMEDIATE(/obj/effect/mapping_helpers/no_lava)
+
+		if(istype(current_thing, /obj/machinery/light) && prob(40))
+			var/obj/machinery/light/light_to_smash = current_thing
+			light_to_smash.break_light_tube(TRUE)
```
Named parameter

Mothblocks commented on code/modules/antagonists/revolution/revolution.dm:
```diff
@@ -443,15 +443,25 @@
 			"target" = get_turf(real_headrev),
 			"style" = STYLE_SYNDICATE,
-			"spawn" = /obj/item/station_charter/revolution,
+			"spawn" = headrev_charter_goodies,
```
Inline the list here instead.

Mothblocks commented on code/modules/antagonists/revolution/revolution.dm:
```diff
@@ -443,15 +443,25 @@
 			continue
 		charter_given = TRUE
+		var/list/headrev_charter_goodies = list(
+		/obj/item/station_charter/revolution,
```
Indent these up one

Mothblocks commented on code/modules/antagonists/revolution/revolution.dm:
```diff
@@ -443,15 +443,25 @@
+	//Forces the next round to have a station trait that makes command rooms get trashed on roundstart
+	var/list/revolutionary_trait = list(/datum/station_trait/revolutionary_trashing) //Wrapped in a list for json encoding
+
+	rustg_file_write(json_encode(revolutionary_trait), FUTURE_STATION_TRAITS_FILE)
```
Don't do this if the file already exists, since it means that an admin has explicitly set it

the-orange-cow commented:
In the even of a rev loss, every surviving member of command should be given a Chain of Command whip, and everyone gets Nanotrasen bedsheets.

Rhials committed:
inline fixes

Rhials committed:
better file foolery

renames admin forced event .json to "forced station traits file" and uses "future station traits file" for the game-forced ones

kills the mapping helper, migrates it to the station trait

is this ultimately faster than the previous functionality? God, I sure hope so!

Rhials committed:
ditches the helpers (map time this time)

Mothblocks commented on code/datums/station_traits/negative_traits.dm:
```diff
@@ -290,3 +290,95 @@
+	INVOKE_ASYNC(src, PROC_REF(trash_this_place)) //Must be called asynchronously
+
+/datum/station_trait/revolutionary_trashing/proc/trash_this_place()
+	for(var/area/station/command/area_to_trash as anything in GLOB.areas)
+		if(!istype(area_to_trash, /area/station/command)) //Skip if we're not a command area
+			continue
```
```suggestion
	for(var/area/station/command/area_to_trash in GLOB.areas)
```

Rhials committed:
an unimaginably idiotic mistake

KittyNoodle commented:
Will this break the command consoles, because if so there is a chance of instant shuttle call.

optimumtact commented:
There shouldn't be any carry over from previous rev rounds unless rev goes back to ending the round.

Otherwise what's teh whole point of stupid post rev anyway

Farquaar commented:
Sounds kinda cool, but probably shouldn't be a thing after *every* successful rev round.

Rhials committed:
smites the mapping helper sprite

Rhials committed:
Merge remote-tracking branch 'upstream/master' into revolutionary-bedsheets

Wallemations commented:
I second that this should at most be a rare occurrence, though the idea is interesting.
Why not just make it a rare negative station trait? That way it'd still be affected by the whole weight system & all that.

san7890 commented:
> Why not just make it a rare negative station trait? That way it'd still be affected by the whole weight system & all that.

i would be supportive of a PR that puts you in the "dark timeline", regardless of bearing from any previous rounds

Rhials committed:
Merge remote-tracking branch 'upstream/master' into revolutionary-bedsheets

Rhials committed:
snip snip

Rhials committed:
scrape scrape

Rhials committed:
one less file is tocuhed now

Rhials committed:
oops

san7890 commented on code/datums/station_traits/negative_traits.dm:
```diff
@@ -290,3 +290,93 @@
+/datum/station_trait/revolutionary_trashing
+	name = "Post-Revolutionary Fervor"
+	show_in_report = TRUE
+	report_message = "Your station was recently reclaimed from a revolutionary commune. We aren't cleaning up after them."
```
```suggestion
	report_message = "Your station was recently reclaimed from a revolutionary commune. We couldn't clean up after them in time."
```

ehhhh the idea of nanotrasen being so evil that they just let the shift after them deal with it is alright, but i'd much rather the concept of something like this happening out of incompetence rather than some level of malice

Rhials committed:
handles reviews

san7890 commented:
![image](https://user-images.githubusercontent.com/34697715/224465142-10de8d5e-5b89-4a9d-bf2b-bf4533be59bc.png)

the mystery of the disappearing red circle

san7890 commented:
![image](https://user-images.githubusercontent.com/34697715/224465157-87d2cfaf-b15a-4ad9-8e61-fc3421efcb8e.png)

also stray pixel snuck itself in somehow

Rhials committed:
Merge remote-tracking branch 'upstream/master' into revolutionary-bedsheets

Rhials committed:
sprite fixes
